File: src/seogeo/fix.py
from __future__ import annotations

"""Safe autofixes for reproducible SEO/GEO artifacts."""

from pathlib import Path
import re

from seogeo.config import Config
from seogeo.generate import build_link_suggestions, render_llms_txt, render_robots_txt
from seogeo.site import Page, Site, load_site


INTERNAL_HTML_LINK_RE = re.compile(r"(\]\()(/?[^)\s]+?)\.html(\))")
FEATURE_COUNT_RE = re.compile(r"(\d+)\s+features\s+across\s+(\d+)\s+categories", re.IGNORECASE)
FEATURE_PAGE_HEADER_RE = re.compile(r"(##\s+Feature Pages\s*\()(\d+)", re.IGNORECASE)
HEAD_CLOSE_RE = re.compile(r"</head>", re.IGNORECASE)
CANONICAL_RE = re.compile(r'<link\b[^>]*\brel=["\']canonical["\'][^>]*>', re.IGNORECASE)
OG_TITLE_RE = re.compile(r'<meta\b[^>]*\bproperty=["\']og:title["\'][^>]*>', re.IGNORECASE)
OG_DESCRIPTION_RE = re.compile(r'<meta\b[^>]*\bproperty=["\']og:description["\'][^>]*>', re.IGNORECASE)
OG_TYPE_RE = re.compile(r'<meta\b[^>]*\bproperty=["\']og:type["\'][^>]*>', re.IGNORECASE)
OG_URL_RE = re.compile(r'<meta\b[^>]*\bproperty=["\']og:url["\'][^>]*>', re.IGNORECASE)
TWITTER_CARD_RE = re.compile(r'<meta\b[^>]*\bname=["\']twitter:card["\'][^>]*>', re.IGNORECASE)
BODY_CLOSE_RE = re.compile(r"</body>", re.IGNORECASE)
ANCHOR_RE = re.compile(r"<a\b[^>]*>.*?</a>", re.IGNORECASE | re.DOTALL)


def _render_related_links_section(source_route: str, target_routes: list[str], heading: str) -> str:
    items = "\n".join(
        f'      <li><a href="/{target}">{target.replace("-", " ").replace("/", " / ").title()}</a></li>'
        for target in target_routes
    )
    return (
        f'\n  <section data-ui="related-links" data-source-route="{_escape_html_attr(source_route or "/")}">\n'
        f"    <h2>{_escape_html_attr(heading)}</h2>\n"
        f"    <ul>\n{items}\n    </ul>\n"
        f"  </section>\n"
    )


def _candidate_anchor_phrases(target_route: str, site: Site) -> list[str]:
    target_page = site.route_pages.get(target_route)
    phrases = []
    if target_page and target_page.title:
        phrases.append(target_page.title)
    phrases.append(target_route.split("/")[-1].replace("-", " "))
    return [phrase for phrase in phrases if phrase]


def _insert_inline_link(raw_text: str, target_route: str, site: Site) -> str:
    """Attempt to insert a natural inline link before falling back to a related-links block."""
    body_start = raw_text.find("<body")
    if body_start == -1:
        return raw_text
    body_content_start = raw_text.find(">", body_start)
    if body_content_start == -1:
        return raw_text
    protected_spans = [match.span() for match in ANCHOR_RE.finditer(raw_text)]
    for phrase in _candidate_anchor_phrases(target_route, site):
        search_index = raw_text.lower().find(phrase.lower(), body_content_start)
        while search_index != -1:
            if not any(start <= search_index < end for start, end in protected_spans):
                replacement = f'<a href="/{target_route}">{raw_text[search_index:search_index + len(phrase)]}</a>'
                return raw_text[:search_index] + replacement + raw_text[search_index + len(phrase) :]
            search_index = raw_text.lower().find(phrase.lower(), search_index + len(phrase))
    return raw_text


def _escape_html_attr(value: str) -> str:
    return (
        value.replace("&", "&amp;")
        .replace('"', "&quot;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
    )


def _canonical_url_for_page(route: str, site_url: str) -> str:
    normalized_site = site_url.rstrip("/")
    if not route:
        return f"{normalized_site}/"
    return f"{normalized_site}/{route}"


def _render_missing_head_tags(page: Page, config: Config) -> list[str]:
    tags: list[str] = []
    if config.site_url and page.canonical is None:
        tags.append(f'<link rel="canonical" href="{_escape_html_attr(_canonical_url_for_page(page.route, config.site_url))}">')
    if "og:title" not in page.metadata and page.title:
        tags.append(f'<meta property="og:title" content="{_escape_html_attr(page.title)}">')
    if "og:description" not in page.metadata and page.meta_description:
        tags.append(f'<meta property="og:description" content="{_escape_html_attr(page.meta_description)}">')
    if "og:type" not in page.metadata:
        tags.append('<meta property="og:type" content="website">')
    if "twitter:card" not in page.metadata and config.default_twitter_card:
        tags.append(f'<meta name="twitter:card" content="{_escape_html_attr(config.default_twitter_card)}">')
    if config.site_url and "og:url" not in page.metadata:
        tags.append(f'<meta property="og:url" content="{_escape_html_attr(_canonical_url_for_page(page.route, config.site_url))}">')
    return tags


def _inject_head_tags(raw_text: str, tags: list[str]) -> str:
    if not tags:
        return raw_text
    insertion = "  " + "\n  ".join(tags) + "\n"
    if HEAD_CLOSE_RE.search(raw_text):
        return HEAD_CLOSE_RE.sub(insertion + "</head>", raw_text, count=1)
    return raw_text


def _apply_html_metadata_fixes(site: Site, config: Config) -> list[Path]:
    changed: list[Path] = []
    for page in site.route_pages.values():
        raw_text = page.path.read_text()
        updated = _inject_head_tags(raw_text, _render_missing_head_tags(page, config))
        if updated != raw_text:
            page.path.write_text(updated)
            changed.append(page.path)
    return changed


def _apply_related_link_insertions(site: Site, config: Config) -> list[Path]:
    if not config.enable_link_autofix:
        return []
    changed: list[Path] = []
    suggestions = build_link_suggestions(site, top_n=config.link_suggestion_count)
    for source_route, target_routes in suggestions.items():
        source_page = site.route_pages[source_route]
        updated = source_page.raw_text
        for target_route in target_routes:
            updated = _insert_inline_link(updated, target_route, site)
        if updated == source_page.raw_text and 'data-ui="related-links"' not in source_page.raw_text and config.related_links_heading not in source_page.raw_text:
            section = _render_related_links_section(source_route, target_routes, config.related_links_heading)
            if BODY_CLOSE_RE.search(source_page.raw_text):
                updated = BODY_CLOSE_RE.sub(section + "</body>", source_page.raw_text, count=1)
            else:
                updated = source_page.raw_text + section
        if updated != source_page.raw_text:
            source_page.path.write_text(updated)
            changed.append(source_page.path)
    return changed


def normalize_internal_llms_links(text: str) -> str:
    """Rewrite internal ``.html`` links in ``llms.txt`` to clean routes."""
    return INTERNAL_HTML_LINK_RE.sub(r"\1\2\3", text).replace("/index)", "/)")


def update_llms_claims(text: str, expected_feature_count: int, expected_category_count: int | None) -> str:
    """Rewrite derived feature counts in ``llms.txt`` when present."""
    updated = text
    if expected_category_count is not None:
        updated = FEATURE_COUNT_RE.sub(
            f"{expected_feature_count} features across {expected_category_count} categories",
            updated,
            count=1,
        )
    updated = FEATURE_PAGE_HEADER_RE.sub(rf"\g<1>{expected_feature_count}", updated, count=1)
    return updated


def apply_safe_fixes(root: Path, config: Config) -> list[Path]:
    """Apply deterministic low-risk fixes to common SEO/GEO artifacts."""
    changed: list[Path] = []
    site = load_site(root, config)
    llms_path = root / "llms.txt"
    if site.llms_text is not None:
        generated = render_llms_txt(site, config.site_url)
        fixed = normalize_internal_llms_links(site.llms_text)
        feature_count = generated.count("](/features/")
        category_line = next((line for line in generated.splitlines() if "features across" in line), None)
        category_count = None
        if category_line:
            match = FEATURE_COUNT_RE.search(category_line)
            if match:
                category_count = int(match.group(2))
        fixed = update_llms_claims(fixed, feature_count, category_count)
        if fixed != site.llms_text:
            llms_path.write_text(fixed)
            changed.append(llms_path)

    robots_path = root / "robots.txt"
    if config.site_url and site.robots_text is None:
        robots_path.write_text(render_robots_txt(config.site_url))
        changed.append(robots_path)
    elif config.site_url and site.robots_text is not None and "sitemap:" not in site.robots_text.lower():
        text = site.robots_text.rstrip() + f"\nSitemap: {config.site_url.rstrip('/')}/sitemap.xml\n"
        robots_path.write_text(text)
        changed.append(robots_path)
    refreshed_site = load_site(root, config)
    changed.extend(_apply_html_metadata_fixes(refreshed_site, config))
    changed.extend(_apply_related_link_insertions(load_site(root, config), config))
    return sorted(set(changed))

File: crates/seogeo-core/src/fix.rs
use anyhow::Result;
use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf};

use crate::config::Config;
use crate::generate::{build_link_suggestions, render_llms_txt, render_robots_txt};
use crate::site::{Page, Site, load_site};

fn escape_html_attr(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('"', "&quot;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

fn canonical_url_for_page(route: &str, site_url: &str) -> String {
    let normalized_site = site_url.trim_end_matches('/');
    if route.is_empty() {
        format!("{}/", normalized_site)
    } else {
        format!("{}/{}", normalized_site, route)
    }
}

fn render_missing_head_tags(page: &Page, config: &Config) -> Vec<String> {
    let mut tags = Vec::new();
    if let Some(site_url) = config.site_url.as_deref() {
        if page.canonical.is_none() {
            tags.push(format!(
                "<link rel=\"canonical\" href=\"{}\">",
                escape_html_attr(&canonical_url_for_page(&page.route, site_url))
            ));
        }
        if page.metadata("og:url").is_none() {
            tags.push(format!(
                "<meta property=\"og:url\" content=\"{}\">",
                escape_html_attr(&canonical_url_for_page(&page.route, site_url))
            ));
        }
    }
    if page.metadata("og:title").is_none()
        && let Some(title) = &page.title
    {
        tags.push(format!(
            "<meta property=\"og:title\" content=\"{}\">",
            escape_html_attr(title)
        ));
    }
    if page.metadata("og:description").is_none()
        && let Some(description) = page.meta_description()
    {
        tags.push(format!(
            "<meta property=\"og:description\" content=\"{}\">",
            escape_html_attr(description)
        ));
    }
    if page.metadata("og:type").is_none() {
        tags.push("<meta property=\"og:type\" content=\"website\">".to_string());
    }
    if page.metadata("twitter:card").is_none() && !config.default_twitter_card.is_empty() {
        tags.push(format!(
            "<meta name=\"twitter:card\" content=\"{}\">",
            escape_html_attr(&config.default_twitter_card)
        ));
    }
    tags
}

fn inject_head_tags(raw_text: &str, tags: &[String]) -> String {
    if tags.is_empty() {
        return raw_text.to_string();
    }
    let insertion = format!("  {}\n", tags.join("\n  "));
    if let Some(index) = raw_text.to_ascii_lowercase().find("</head>") {
        let mut updated = String::with_capacity(raw_text.len() + insertion.len());
        updated.push_str(&raw_text[..index]);
        updated.push_str(&insertion);
        updated.push_str(&raw_text[index..]);
        return updated;
    }
    raw_text.to_string()
}

fn apply_html_metadata_fixes(site: &Site, config: &Config) -> Result<Vec<PathBuf>> {
    let mut changed = Vec::new();
    for page in site.route_pages() {
        let raw_text = fs::read_to_string(&page.path)?;
        let updated = inject_head_tags(&raw_text, &render_missing_head_tags(page, config));
        if updated != raw_text {
            fs::write(&page.path, updated)?;
            changed.push(page.path.clone());
        }
    }
    Ok(changed)
}

fn render_related_links_section(
    source_route: &str,
    target_routes: &[String],
    heading: &str,
) -> String {
    let items = target_routes
        .iter()
        .map(|target| {
            let label = target
                .replace('-', " ")
                .replace('/', " / ")
                .split_whitespace()
                .map(|part| {
                    let mut chars = part.chars();
                    match chars.next() {
                        Some(first) => first.to_ascii_uppercase().to_string() + chars.as_str(),
                        None => String::new(),
                    }
                })
                .collect::<Vec<_>>()
                .join(" ");
            format!("      <li><a href=\"/{}\">{}</a></li>", target, label)
        })
        .collect::<Vec<_>>()
        .join("\n");
    format!(
        "\n  <section data-ui=\"related-links\" data-source-route=\"{}\">\n    <h2>{}</h2>\n    <ul>\n{}\n    </ul>\n  </section>\n",
        escape_html_attr(if source_route.is_empty() {
            "/"
        } else {
            source_route
        }),
        escape_html_attr(heading),
        items,
    )
}

fn apply_related_link_insertions(site: &Site, config: &Config) -> Result<Vec<PathBuf>> {
    if !config.enable_link_autofix {
        return Ok(Vec::new());
    }
    let suggestions = build_link_suggestions(site, config.link_suggestion_count);
    let mut changed = Vec::new();
    for (source_route, target_routes) in suggestions {
        let Some(source_page) = site.page(&source_route) else {
            continue;
        };
        let raw_text = fs::read_to_string(&source_page.path)?;
        if raw_text.contains("data-ui=\"related-links\"")
            || raw_text.contains(&config.related_links_heading)
        {
            continue;
        }
        let section = render_related_links_section(
            &source_route,
            &target_routes,
            &config.related_links_heading,
        );
        let updated = if let Some(index) = raw_text.to_ascii_lowercase().find("</body>") {
            let mut text = String::with_capacity(raw_text.len() + section.len());
            text.push_str(&raw_text[..index]);
            text.push_str(&section);
            text.push_str(&raw_text[index..]);
            text
        } else {
            format!("{}{}", raw_text, section)
        };
        if updated != raw_text {
            fs::write(&source_page.path, updated)?;
            changed.push(source_page.path.clone());
        }
    }
    Ok(changed)
}

pub fn apply_safe_fixes(root: &Path, config: &Config) -> Result<Vec<PathBuf>> {
    let mut changed = BTreeSet::new();
    let site = load_site(root)?;

    let llms_path = root.join("llms.txt");
    if site.llms_text.is_some() {
        let generated = render_llms_txt(&site, config.site_url.as_deref());
        if site.llms_text.as_deref() != Some(generated.as_str()) {
            fs::write(&llms_path, generated)?;
            changed.insert(llms_path);
        }
    }

    if let Some(site_url) = config.site_url.as_deref() {
        let robots_path = root.join("robots.txt");
        match site.robots_text.as_deref() {
            None => {
                fs::write(&robots_path, render_robots_txt(site_url))?;
                changed.insert(robots_path);
            }
            Some(text) if !text.to_ascii_lowercase().contains("sitemap:") => {
                let updated = format!(
                    "{}\nSitemap: {}/sitemap.xml\n",
                    text.trim_end(),
                    site_url.trim_end_matches('/')
                );
                fs::write(&robots_path, updated)?;
                changed.insert(robots_path);
            }
            _ => {}
        }
    }

    let refreshed_site = load_site(root)?;
    for path in apply_html_metadata_fixes(&refreshed_site, config)? {
        changed.insert(path);
    }
    let refreshed_site = load_site(root)?;
    for path in apply_related_link_insertions(&refreshed_site, config)? {
        changed.insert(path);
    }

    Ok(changed.into_iter().collect())
}

#[cfg(test)]
mod tests {
    use super::apply_safe_fixes;
    use crate::config::Config;
    use anyhow::Result;
    use std::fs;

    fn write(path: &std::path::Path, text: &str) {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).unwrap();
        }
        fs::write(path, text).unwrap();
    }

    fn make_html_page(route: &str, body: &str) -> String {
        let canonical = if route.is_empty() {
            "https://example.com/".to_string()
        } else {
            format!("https://example.com/{}", route)
        };
        format!(
            "<html lang=\"en\"><head><title>x</title><meta name=\"description\" content=\"y\"><link rel=\"canonical\" href=\"{}\"></head><body><h1>x</h1>{}</body></html>",
            canonical, body,
        )
    }

    #[test]
    fn updates_llms_and_creates_robots_and_head_tags() -> Result<()> {
        let temp_dir = tempfile::tempdir().unwrap();
        let root = temp_dir.path();
        write(
            &root.join("index.html"),
            "<html><head><title>x</title><meta name=\"description\" content=\"y\"></head><body><h1>x</h1></body></html>",
        );
        write(
            &root.join("features/alpha/index.html"),
            &make_html_page("features/alpha", ""),
        );
        write(
            &root.join("feature-data.json"),
            r#"{"categories":[{"id":"x","name":"X","features":[{"slug":"alpha"}]}]}"#,
        );
        write(
            &root.join("llms.txt"),
            "# Site\n\n## Key Facts\n- 9 features across 3 categories\n\n## Pages\n- [Home](index.html)\n",
        );
        let config = Config {
            site_url: Some("https://example.com".to_string()),
            ..Config::default()
        };
        let changed = apply_safe_fixes(root, &config)?;
        assert!(!changed.is_empty());
        assert!((root.join("robots.txt")).exists());
        let index_html = fs::read_to_string(root.join("index.html"))?;
        assert!(index_html.contains("rel=\"canonical\""));
        assert!(index_html.contains("property=\"og:title\""));
        assert!(index_html.contains("name=\"twitter:card\""));
        assert!(fs::read_to_string(root.join("llms.txt"))?.contains("features/alpha)"));
        Ok(())
    }

    #[test]
    fn can_insert_related_links_when_enabled() -> Result<()> {
        let temp_dir = tempfile::tempdir().unwrap();
        let root = temp_dir.path();
        write(
            &root.join("index.html"),
            &make_html_page("", "<p>Guide to alpha workflows and alpha setup.</p>"),
        );
        write(
            &root.join("guide/index.html"),
            &make_html_page("guide", "<p>Guide body alpha workflows.</p>"),
        );
        write(
            &root.join("alpha/index.html"),
            &make_html_page("alpha", "<p>Alpha workflows reference.</p>"),
        );
        let config = Config {
            site_url: Some("https://example.com".to_string()),
            enable_link_autofix: true,
            related_links_heading: "Related pages".to_string(),
            ..Config::default()
        };
        apply_safe_fixes(root, &config)?;
        let updated = fs::read_to_string(root.join("index.html"))?;
        assert!(updated.contains("/alpha") || updated.contains("data-ui=\"related-links\""));
        Ok(())
    }
}
